compiler/src/lnntoamm/Statement.ts

```diff
@@ -0,0 +1,785 @@
+          let callAst = new NamedAnd(
+            work.get('variable').t + next.get('fncall').t,
+            {
+              fnname: work.get('variable'),
+              fncall: next.get('fncall'),
+            },
+            (work as NamedOr).filename,
+            work.line,
+            work.char,
+          );
+          stmts.push(...Call.fromAsts(
+            callAst,
+            // null,
+            varName,
+            next.get('fncall'),
+            metadata,
+          ));
+          ii += 1;
+        } else if (next.has('methodsep')) {
+          TODO('accesses/methods');
+        } else {
+          throw new Error(`unexpected token: expected dot or call, found ${next}`);
+        }
+      } else if (work.has('constants')) {
+        work = work.get('constants');
+        let dec = Dec.generate(Lit.fromConstantsAst(work, metadata));
+        metadata.define(dec);
+        stmts.push(dec);
+      } else {
+        // TODO: don't lump in HOF and chains
+        throw new Error(`unexpected token: expected variable or value, found ${work.t.trim()}`);
+      }
+    }
+
+    return stmts;
+  }
+}
+
+export class Assign extends Stmt {
+  upstream: Dec
+  val: Stmt
+
+  constructor(
+    ast: LPNode,
+    upstream: Dec,
+    assignTo: Stmt,
+  ) {
+    super(ast);
+    this.upstream = upstream;
+    this.val = assignTo;
+  }
+
+  static fromAssignmentsAst(ast: LPNode, metadata: MetaData): Stmt[] {
+    let stmts: Stmt[] = [];
+    const name = ast.get('varn').t;
+    const upstream = metadata.var(name);
+    if (upstream === null) {
+      throw new Error(`can't assign to ${name}: not found`);
+    }
+    stmts.push(...Stmt.fromAssignables(ast.get('assignables'), metadata));
+    const expr = stmts.pop();
+    let assign: Assign;
+    if (expr instanceof Dec) {
+      assign = new Assign(ast, upstream, expr.val);
+    } else if (expr instanceof Ref) {
+      TODO('figure out how to resolve this???');
+    } else {
+      throw new Error(`invalid assignment state: not a declaration`);
+    }
+    upstream.ty.constrain(assign.val.exprTy());
+    stmts.push(assign);
+    return stmts;
+  }
+
+  exprTy(): Type {
+    throw new Error(`assignments aren't expressions`);
+  }
+
+  inline(amm: Output) {
+    const name = this.upstream.ammName;
+    const ty = this.upstream.ty.breakdown(); // always use the declaration's type, since it's been reduced.
+    if (this.val instanceof Call) {
+      this.val.inline(amm, name, ty, '');
+    } else if (this.val instanceof Lit) {
+      amm.assign('', name, ty, this.val.val);
+    } else {
+      throw new Error(`Unexpected assignment expression: ${this.val}`);
+    }
+  }
+}
+
+export class Call extends Stmt {
+  fn: Fn
+  args: Ref[]
+
+  constructor(
+    ast: LPNode,
+    fn: Fn,
+    args: Ref[],
+  ) {
+    super(ast)
+    this.fn = fn;
+    this.args = args;
+  }
+
+  static fromAsts(
+    wholeAst: LPNode,
+    fnName: string,
+    fnCallAst: LPNode,
```

Comment:
Why does this exist? Why not just extract the `fnCallAst` from `wholeAst`?

Reply:
`wholeAst` is only used for eventual error reporting - see lines 100-109

Reply:
I just mean it's de-centralizing extraction of the function call logic, so it doesn't seem DRY?